File: spring boot/MicroservicioProductos/src/main/java/controller/ProductController.java
package controller;

import model.Product;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.ArrayList;
import java.util.List;

@RestController
@RequestMapping("/products")
public class ProductController {

    private List<Product> products = new ArrayList<>(List.of(
            new Product(1, "Laptop", "Portátil potente con 16GB RAM", 1200.99),
            new Product(2, "Smartphone", "Pantalla AMOLED, carga rápida", 799.50),
            new Product(3, "Auriculares", "Cancelación de ruido", 199.99)
    ));

    // Obtener todos los productos
    @GetMapping
    public ResponseEntity<List<Product>> getAllProducts() {
        return ResponseEntity.ok(products);
    }

    // Obtener producto por ID
    @GetMapping("/{id}")
    public ResponseEntity<Product> getProductById(@PathVariable Integer id) {
        return products.stream()
                .filter(p -> p.getId().equals(id))  // <-- Aquí está la búsqueda
                .findFirst()
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

 // Crear producto (POST)
    @PostMapping
    public ResponseEntity<String> createProduct(@RequestBody Product product) {
        products.add(product);
        return ResponseEntity.ok("Producto '" + product.getName() + "' creado exitosamente.");
    }

    // Actualizar producto (PUT)
    @PutMapping("/{id}")
    public ResponseEntity<String> updateProduct(@PathVariable Integer id, @RequestBody Product product) {
        for (Product p : products) {
            if (p.getId().equals(id)) {
                p = new Product(id, product.getName(), product.getDescription(), product.getPrice());
                return ResponseEntity.ok("Producto actualizado correctamente.");
            }
        }
        return ResponseEntity.notFound().build();
    }

    // Eliminar producto
    @DeleteMapping("/{id}")
    public ResponseEntity<String> deleteProduct(@PathVariable Integer id) {
        boolean removed = products.removeIf(p -> p.getId().equals(id));
        return removed ? ResponseEntity.ok("Producto eliminado correctamente.") : ResponseEntity.notFound().build();
    }

}

File: microservice-laravel-ud6/app/Http/Controllers/ProductController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use Illuminate\Support\Facades\Http;

class ProductController extends Controller
{
    // Obtener lista de productos
    public function fetchProducts()
    {
        try {
            $response = Http::get('http://localhost:8081/products');

            if ($response->successful()) {
                return response()->json($response->json(), 200);
            } else {
                return response()->json([
                    'error' => 'Error en el microservicio',
                    'status' => $response->status()
                ], $response->status());
            }
        } catch (\Exception $e) {
            return response()->json([
                'error' => 'No se pudo conectar al microservicio',
                'message' => $e->getMessage()
            ], 500);
        }
    }


    // Obtener un producto por ID
    public function fetchProductById($id)
    {
        try {
            $response = Http::get('http://localhost:8081/products/' . $id);

            if ($response->successful()) {
                return response()->json($response->json(), 200);
            } else {
                return response()->json(['error' => 'Producto no encontrado', 'status' => $response->status()], $response->status());
            }
        } catch (\Exception $e) {
            return response()->json(['error' => 'No se pudo conectar al microservicio', 'message' => $e->getMessage()], 500);
        }
    }

    // Crear producto (POST)
    public function createProduct(Request $request)
    {
        try {
            $response = Http::post('http://localhost:8081/products', $request->all());

            if ($response->successful()) {
                return response()->json(['message' => 'Producto creado correctamente'], 201);
            } else {
                return response()->json(['error' => 'Error al crear el producto'], $response->status());
            }
        } catch (\Exception $e) {
            return response()->json(['error' => 'No se pudo conectar al microservicio', 'message' => $e->getMessage()], 500);
        }
    }

    // Actualizar producto (PUT)
    public function updateProduct(Request $request, $id)
    {
        try {
            $response = Http::put('http://localhost:8081/products/' . $id, $request->all());

            if ($response->successful()) {
                return response()->json(['message' => 'Producto actualizado correctamente'], 200);
            } else {
                return response()->json(['error' => 'Error al actualizar el producto'], $response->status());
            }
        } catch (\Exception $e) {
            return response()->json(['error' => 'No se pudo conectar al microservicio', 'message' => $e->getMessage()], 500);
        }
    }

    // Eliminar producto (DELETE)
    public function deleteProduct($id)
    {
        try {
            $response = Http::delete('http://localhost:8081/products/' . $id);

            if ($response->successful()) {
                return response()->json(['message' => 'Producto eliminado correctamente'], 200);
            } else {
                return response()->json(['error' => 'Error al eliminar el producto'], $response->status());
            }
        } catch (\Exception $e) {
            return response()->json(['error' => 'No se pudo conectar al microservicio', 'message' => $e->getMessage()], 500);
        }
    }
}
